File: mobile/kmp/ui/src/commonMain/kotlin/com/atls/hyperion/ui/components/button/styles/shape/Variants.kt
package com.atls.hyperion.ui.components.button.styles.shape

import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.runtime.Composable
import com.atls.hyperion.ui.theme.LocalHyperionElevations
import com.atls.hyperion.ui.theme.LocalHyperionTypography
import com.atls.hyperion.ui.theme.tokens.borders.BorderWidth
import com.atls.hyperion.ui.theme.tokens.layout.Radii
import com.atls.hyperion.ui.theme.tokens.layout.Spacing

@Composable
fun ButtonShape.Companion.xs(): ButtonShape =
    ButtonShape(
        minHeight = MinHeight.xs,
        cornerRadius = Radii.xs2,
        paddings = PaddingValues(
            horizontal = Spacing.component.sm,
            vertical = Spacing.component.xs
        ),
        typography = LocalHyperionTypography.current.label.sm,
        borderWidth = BorderWidth.sm,
        gap = Spacing.component.xs,
        addonSize = Spacing.component.lg,
        elevations = LocalHyperionElevations.current.xs
    )

@Composable
fun ButtonShape.Companion.sm(): ButtonShape =
    ButtonShape(
        minHeight = MinHeight.sm,
        cornerRadius = Radii.xs2,
        paddings = PaddingValues(
            horizontal = Spacing.component.md,
            vertical = Spacing.component.sm
        ),
        typography = LocalHyperionTypography.current.body.md,
        borderWidth = BorderWidth.sm,
        gap = Spacing.component.md,
        addonSize = Spacing.component.xl,
        elevations = LocalHyperionElevations.current.sm
    )

@Composable
fun ButtonShape.Companion.md(): ButtonShape =
    ButtonShape(
        minHeight = MinHeight.md,
        cornerRadius = Radii.xs2,
        paddings = PaddingValues(
            horizontal = Spacing.component.lg,
            vertical = Spacing.component.md
        ),
        typography = LocalHyperionTypography.current.label.md,
        borderWidth = BorderWidth.sm,
        gap = Spacing.component.lg,
        addonSize = Spacing.component.xl2,
        elevations = LocalHyperionElevations.current.md
    )

@Composable
fun ButtonShape.Companion.lg(): ButtonShape =
    ButtonShape(
        minHeight = MinHeight.lg,
        cornerRadius = Radii.xs2,
        paddings = PaddingValues(
            horizontal = Spacing.component.xl,
            vertical = Spacing.component.sm
        ),
        typography = LocalHyperionTypography.current.label.lg,
        borderWidth = BorderWidth.sm,
        gap = Spacing.component.xl,
        addonSize = Spacing.component.xl2,
        elevations = LocalHyperionElevations.current.lg
    )

File: ui-parts/button/src/styles/shape/variants.ts
import type { Theme }                 from '@atls-ui/theme'

import type { ButtonShape }           from './interfaces.js'
import type { ButtonShapeName }       from './interfaces.js'
import type { ButtonShapeProperties } from './interfaces.js'
import type { ButtonShapes }          from './interfaces.js'

import { borderStyles }               from '@atls-ui/theme/tokens'
import { borderWidths }               from '@atls-ui/theme/tokens'
import { radii }                      from '@atls-ui/theme/tokens'
import { spacing }                    from '@atls-ui/theme/tokens'

import { buttonAddonSizes }           from './constants.js'

const variants = (theme: Theme): Record<ButtonShapeName, ButtonShapeProperties> => ({
  xs: {
    borderRadius: radii.xs2,
    paddingInline: spacing.component.sm,
    paddingBlock: spacing.component.xs,
    typography: theme.typography.label.sm,
    borderStyle: borderStyles.solid,
    borderWidth: borderWidths.sm,
    gap: spacing.component.xs,
    addonSize: buttonAddonSizes.xs,
    elevations: {
      ...theme.elevations.xs,
      focused: theme.elevations.xs.disabled,
    },
  },
  sm: {
    borderRadius: radii.xs2,
    paddingInline: spacing.component.md,
    paddingBlock: spacing.component.xs,
    typography: theme.typography.body.md,
    borderStyle: borderStyles.solid,
    borderWidth: borderWidths.sm,
    gap: spacing.component.md,
    addonSize: buttonAddonSizes.sm,
    elevations: {
      ...theme.elevations.sm,
      focused: theme.elevations.sm.disabled,
    },
  },
  md: {
    borderRadius: radii.xs2,
    paddingInline: spacing.component.lg,
    paddingBlock: spacing.component.sm,
    typography: theme.typography.label.md,
    borderStyle: borderStyles.solid,
    borderWidth: borderWidths.sm,
    gap: spacing.component.lg,
    addonSize: buttonAddonSizes.md,
    elevations: {
      ...theme.elevations.md,
      focused: theme.elevations.md.disabled,
    },
  },
  lg: {
    borderRadius: radii.xs2,
    paddingInline: spacing.component.xl,
    paddingBlock: spacing.component.sm,
    typography: theme.typography.label.lg,
    borderStyle: borderStyles.solid,
    borderWidth: borderWidths.sm,
    gap: spacing.component.xl,
    addonSize: buttonAddonSizes.lg,
    elevations: {
      ...theme.elevations.lg,
      focused: theme.elevations.lg.disabled,
    },
  },
})

export const buttonShapes = {
  xs: 'xs',
  sm: 'sm',
  md: 'md',
  lg: 'lg',
} as const satisfies ButtonShapes

export const isButtonShapeName = (shape: string): shape is ButtonShapeName => shape in buttonShapes

export const resolveButtonShape = (
  shape: ButtonShape,
  theme: Theme
): ButtonShapeProperties | undefined => {
  if (typeof shape !== 'string') {
    return shape
  }

  return isButtonShapeName(shape) ? variants(theme)[shape] : undefined
}
